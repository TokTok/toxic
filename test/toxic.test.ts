import { test, expect, Shell } from "@microsoft/tui-test";

test.use({ shell: Shell.Bash, rows: 25, columns: 90 });

const fast = { timeout: 100 };
const slow = { timeout: 1000 };

async function startToxic(terminal: Shell.Terminal) {
  terminal.write("toxic\r");
  return expect(terminal.getByText("Welcome to Toxic")).toBeVisible(slow);
}

test.describe("toxic", () => {
  test("toxic --help shows usage message", async ({ terminal }) => {
    terminal.write("toxic --help\r");
    await expect(terminal.getByText("usage: toxic")).toBeVisible(fast);
    await expect(terminal).toMatchSnapshot();
  });

  test("toxic shows intro message", async ({ terminal }) => {
    terminal.write("toxic\r");
    await expect(
      terminal.getByText("Would you like to encrypt it")
    ).toBeVisible(fast);
    terminal.write("n\r");
    await expect(terminal.getByText("Welcome to Toxic")).toBeVisible(slow);
    await expect(terminal).toMatchSnapshot();
  });

  test("help window works", async ({ terminal }) => {
    await startToxic(terminal);
    terminal.write("/help\r");
    await expect(terminal.getByText("Help Menu")).toBeVisible(fast);
    await expect(terminal).toMatchSnapshot();
  });

  test("global help window works", async ({ terminal }) => {
    await startToxic(terminal);
    terminal.write("/help\r");
    await expect(terminal.getByText("Help Menu")).toBeVisible(fast);
    terminal.write("g");
    await expect(terminal.getByText("Global Commands")).toBeVisible(fast);
    await expect(terminal).toMatchSnapshot();
  });

  test("help window can be exited", async ({ terminal }) => {
    await startToxic(terminal);
    terminal.write("/help\r");
    await expect(terminal.getByText("Help Menu")).toBeVisible(fast);
    terminal.write("x");
    await expect(terminal.getByText("Help Menu")).not.toBeVisible(fast);
    await expect(terminal).toMatchSnapshot();
  });

  test("terminal resize works", async ({ terminal }) => {
    await startToxic(terminal);
    terminal.write("/help\r");
    await expect(terminal.getByText("Help Menu")).toBeVisible(fast);
    terminal.resize(60, 20);
    await expect(terminal.getByText("Help Menu")).toBeVisible(fast);
    // Wait for one of the last things to be drawn before screenshotting.
    await expect(terminal.getByText("[Contacts]")).toBeVisible(fast);
    await expect(terminal).toMatchSnapshot();
  });

  test("last message is shown when terminal gets smaller", async ({ terminal }) => {
    await startToxic(terminal);
    // Add a friend.
    terminal.write("/add 13117C65771C5A05409F532A7809D238E38E94312C870FE7970C5B65B1215E20CEF40A1D48B7\r");
    await expect(terminal.getByText("Friend request sent.")).toBeVisible(fast);
    terminal.write(String.fromCharCode(16));
    await expect(terminal.getByText("Press the h key")).toBeVisible(fast);
    terminal.write("\r");
    await expect(terminal.getByText("[Offline]")).toBeVisible(fast);
    for (let i = 1; i <= 10; ++i) {
        terminal.write(`we're (${i}) writing some pretty long message here, this is message ${i}\r`);
    }
    await expect(terminal.getByText("message 10")).toBeVisible(fast);
    terminal.resize(60, 20);
    // slow, because resize can take a while
    // TODO(JFreegman): Make this "message 10" when it's fixed.
    await expect(terminal.getByText("message 7")).toBeVisible(slow);
    await expect(terminal).toMatchSnapshot();
  });
});
