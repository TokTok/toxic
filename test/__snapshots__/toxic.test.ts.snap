// TUI Test Snapshot v1

exports[`global help window works 1`] = String.raw`
╭──────────────────────────────────────────────────────────────────────────────────────────╮
│┌──────────────────────────────────────────────────────────────────────────────┐          │
││Global Commands:                                                              │          │
││ /add <addr> <msg>          : Add contact with optional message               │          │
││ /accept <id>               : Accept friend request                           │          │
││ /avatar <path>             : Set an avatar (leave path empty to unset)       │          │
││ /color <color>             : Change the colour of the focused window's name  │          │
││ /conference <type>         : Create a conference where type: text | audio    │          │
││ /connect <ip> <port> <key> : Manually connect to a DHT node                  │          │
││ /decline <id>              : Decline friend request                          │          │
││ /requests                  : List pending friend requests                    │          │
││ /status <type>             : Set status (Online, Busy, Away)                 │          │
││ /note <msg>                : Set a personal note                             │          │
││ /nick <name>               : Set your global name (doesn't affect groups)    │          │
││ /nospam <value>            : Change part of your Tox ID to stop spam         │          │
││ /log <on>|<off>            : Enable/disable logging                          │          │
││ /myid                      : Print your Tox ID                               │          │
││ /group <name>              : Create a new group chat                         │          │
││ /join <chatid>             : Join a public groupchat using a Chat ID         │          │
││ /game <name>               : Play a game                                     │          │
││ /clear                     : Clear window history                            │          │
││ /close                     : Close the current chat window                   │          │
││ /quit or /exit             : Exit Toxic                                      │          │
││                                                                              │          │
││ main menu | exit                                                             │          │
│└──────────────────────────────────────────────────────────────────────────────┘          │
╰──────────────────────────────────────────────────────────────────────────────────────────╯
`;

exports[`help window can be exited 1`] = String.raw`
╭──────────────────────────────────────────────────────────────────────────────────────────╮
│ [Offline] Toxic User                                                                     │
│    _____ _____  _____ ____                                                               │
│   |_   _/ _ \ \/ /_ _/ ___|                                                              │
│     | || | | \  / | | |                                                                  │
│     | || |_| /  \ | | |___                                                               │
│     |_| \___/_/\_\___\____| v.0.14.1                                                     │
│                                                                                          │
│Welcome to Toxic, a free, open source Tox-based instant messaging client.                 │
│Type "/help" for assistance. Further help may be found via the man page.                  │
│                                                                                          │
│Avatar has been unset.                                                                    │
│$ /help                                                                                   │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│  [Home] [Contacts]                                                                       │
│                                                                                          │
╰──────────────────────────────────────────────────────────────────────────────────────────╯
`;

exports[`help window works 1`] = String.raw`
╭──────────────────────────────────────────────────────────────────────────────────────────╮
│ [Offline] Toxic User                                                                     │
│    _____ _____  _____ ____                                                               │
│   |_   _/ _ \ \/ /_ _/ ___|                                                              │
│   ┌────────────────────────┐                                                             │
│   │       Help Menu        │                                                             │
│   │global commands         │v.0.14.1                                                     │
│   │chat commands           │                                                             │
│Wel│conference commands     │n source Tox-based instant messaging client.                 │
│Typ│groupchat commands      │Further help may be found via the man page.                  │
│   │friendlist controls     │                                                             │
│Ava│key bindings            │                                                             │
│$ /└────────────────────────┘                                                             │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│  [Home] [Contacts]                                                                       │
│                                                                                          │
╰──────────────────────────────────────────────────────────────────────────────────────────╯
`;

exports[`last message is shown when terminal gets smaller 1`] = String.raw`
╭────────────────────────────────────────────────────────────╮
│ [Offline]                                         {13117C} │
│Logging for this window is OFF; type "/log on" to enable.   │
│21:41 - Toxic User: we're (1) writing some pretty long      │
│                    message here, this is message 1 x       │
│21:41 - Toxic User: we're (2) writing some pretty long      │
│                    message here, this is message 2 x       │
│21:41 - Toxic User: we're (3) writing some pretty long      │
│                    message here, this is message 3 x       │
│21:41 - Toxic User: we're (4) writing some pretty long      │
│                    message here, this is message 4 x       │
│21:41 - Toxic User: we're (5) writing some pretty long      │
│                    message here, this is message 5 x       │
│21:41 - Toxic User: we're (6) writing some pretty long      │
│                    message here, this is message 6 x       │
│21:41 - Toxic User: we're (7) writing some pretty long      │
│                    message here, this is message 7 x       │
│21:41 - Toxic User: we're (8) writing some pretty long      │
│                    message here, this is message 8 x       │
│  [Home] [Contacts] []                                      │
│                                                            │
╰────────────────────────────────────────────────────────────╯
`;

exports[`terminal resize works 1`] = String.raw`
╭────────────────────────────────────────────────────────────╮
│ [Offline] Toxic User                                       │
│    _____ _____  _____ ____                                 │
│   |_   _/ _ \ \/ /_ _/ ___|                                │
│   ┌────────────────────────┐                               │
│   │       Help Menu        │                               │
│   │global commands         │v.0.14.1                       │
│   │chat commands           │                               │
│Wel│conference commands     │n source Tox-based instant     │
│mes│groupchat commands      │                               │
│Typ│friendlist controls     │Further help may be found via  │
│the│key bindings            │                               │
│   └────────────────────────┘                               │
│Avatar has been unset.                                      │
│$ /help                                                     │
│                                                            │
│                                                            │
│                                                            │
│                                                            │
│  [Home] [Contacts]                                         │
│                                                            │
╰────────────────────────────────────────────────────────────╯
`;

exports[`toxic --help shows usage message 1`] = String.raw`
╭──────────────────────────────────────────────────────────────────────────────────────────╮
│> toxic --help                                                                            │
│usage: toxic [OPTION] [FILE ...]                                                          │
│  -4, --ipv4               Force IPv4 connection                                          │
│  -b, --debug              Enable stderr for debugging                                    │
│  -c, --config             Use specified config file                                      │
│  -d, --default-locale     Use default POSIX locale                                       │
│  -e, --encrypt-data       Encrypt an unencrypted data file                               │
│  -f, --file               Use specified data file                                        │
│  -h, --help               Show this message and exit                                     │
│  -l, --logging            Enable toxcore logging: Requires [log_path | stderr]           │
│  -L, --no-lan             Disable local discovery                                        │
│  -n, --nodes              Use specified DHTnodes file                                    │
│  -o, --noconnect          Do not connect to the DHT network                              │
│  -p, --SOCKS5-proxy       Use SOCKS5 proxy: Requires [IP] [port]                         │
│  -P, --HTTP-proxy         Use HTTP proxy: Requires [IP] [port]                           │
│  -r, --namelist           Use specified name lookup server list                          │
│  -t, --force-tcp          Force toxic to use a TCP connection (use with proxies)         │
│  -T, --tcp-server         Act as a TCP relay server: Requires [port]                     │
│  -u, --unencrypt-data     Unencrypt an encrypted data file                               │
│  -v, --version            Print the version                                              │
│>                                                                                         │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
╰──────────────────────────────────────────────────────────────────────────────────────────╯
`;

exports[`toxic shows intro message 1`] = String.raw`
╭──────────────────────────────────────────────────────────────────────────────────────────╮
│ [Offline] Toxic User | Toxing on Toxic                                                   │
│    _____ _____  _____ ____                                                               │
│   |_   _/ _ \ \/ /_ _/ ___|                                                              │
│     | || | | \  / | | |                                                                  │
│     | || |_| /  \ | | |___                                                               │
│     |_| \___/_/\_\___\____| v.0.14.1                                                     │
│                                                                                          │
│Welcome to Toxic, a free, open source Tox-based instant messaging client.                 │
│Type "/help" for assistance. Further help may be found via the man page.                  │
│                                                                                          │
│Avatar has been unset.                                                                    │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│                                                                                          │
│  [Home] [Contacts]                                                                       │
│                                                                                          │
╰──────────────────────────────────────────────────────────────────────────────────────────╯
`;

